_posts/2019-07-30-indexing-searching-nuget-with-azure-functions-and-search.markdown

```diff
@@ -0,0 +1,891 @@
+* ...
+
+The NuGet client uses these depending on the scenario. Obviously, it uses search when searching packages, but will use the flat container for package restore as that API is much better suited for quickly resolving package binaries.
+
+### NuGet catalog
+
+One interesting resource in NuGet's V3 API, is the *Catalog*. It is really an event stream of package additions, updates and deletes. And it's chronological, which means that if we keep track of the atest timestamp we received from the cataog, we can use it to continue where we left off on the next run.
+
+Its structure is a tree. After finding the catalog root from [api.nuget.org/v3/index.json](https://api.nuget.org/v3/index.json), we will find:
+
+* Catalog root - [api.nuget.org/v3/catalog0/index.json](https://api.nuget.org/v3/catalog0/index.json)
+  * N catalog pages - [api.nuget.org/v3/catalog0/page0.json](https://api.nuget.org/v3/catalog0/page0.json)
+    * N catalog leafs - [api.nuget.org/v3/catalog0/data/2015.02.01.06.22.45/adam.jsgenerator.1.1.0.json](https://api.nuget.org/v3/catalog0/data/2015.02.01.06.22.45/adam.jsgenerator.1.1.0.json)
+
+The catalog root contains a bunch of links to catalog pages, with the timestamp they were added/updated. We can then traverse each page, and find out URLs to all catalog leafs, which contain the opration type (update or delete), as well as basic package metadata.
+
+Using a timestamp as a cursor, and traversing some JSON. Seams easy enough to get the data we needed in the first place! And it turns out that there is a NuGet package, [NuGet.Services.Metadata.Catalog](https://www.nuget.org/packages/NuGet.Services.Metadata.Catalog) (code [on GitHub](https://github.com/NuGet/NuGet.Services.Metadata)) which already implemented access to the catalog.
+
+> Note: there's [another NuGet package](https://www.nuget.org/packages/NuGet.CatalogReader/) which provides access to the catalog in case you need it. Use what works best for you!
+
+The `NuGet.Services.Metadata.Catalog` package comes with a `CatalogProcessor` class which reads the catalog between a start and end timestamp, and allows plugging in an `ICatalogLeafProcessor` that we can implement:
+
+```  
+Task<bool> ProcessPackageDetailsAsync(PackageDetailsCatalogLeaf leaf);
+Task<bool> ProcessPackageDeleteAsync(PackageDeleteCatalogLeaf leaf);
+```
+
+That seems easy enough! These two methods get called whenever the `CatalogProcessor` processes a leaf from the catalog, and gives us details about add/update, and delete operations.
+
+> Note: in code examples, you will see me using `BatchCatalogProcessor`. This is a [custom cataog processor](https://gist.github.com/maartenba/7b6d6f5f9a96286314e8bbd65435b9f9) that speeds up traversing the catalog. For every page, it only takes the newest leaf of every single package id + version combination instead of every single operation. For indexing packages, all we'd need is the latest state (is a package considered existing, or considered deleted?).
```

Comment:
```suggestion
> Note: in code examples, you will see me using `BatchCatalogProcessor`. This is a [custom catalog processor](https://gist.github.com/maartenba/7b6d6f5f9a96286314e8bbd65435b9f9) that speeds up traversing the catalog. For every page, it only takes the newest leaf of every single package id + version combination instead of every single operation. For indexing packages, all we'd need is the latest state (is a package considered existing or considered deleted?).
```

Link posts to GitHub Gist, is it not in your GitHub repo?